Fix multiple test failures under ASan by forcing use of `sigaction`

JohnoKing opened the pull request:
This commit fixes multiple test failures when the coprocess, io and pty tests are run under ASan. ~Unlike the ksh93v- approach of calling `sigaction` via a `sh_signal` function, this commit repurposes the (otherwise unused) `signal` wrapper function in libast, which already uses `sigaction` for signal handling.~ (**EDIT:** See the comment below for a more accurate explanation as to what the pull request does.)

ASan test failures fixed in this pull request:
```
$ ASAN_OPTIONS='detect_leaks=0' setarch ${ uname -m } -R bin/shtests -u coprocess io pty
#### Regression-testing /home/johno/GitRepos/KornShell/ksh/arch/linux.i386-64/bin/ksh ####
test coprocess(C.UTF-8) begins at 2024-01-27+18:14:24
	coprocess.sh[196]: FAIL: traps when reading from cat coprocess not working
	coprocess.sh[228]: FAIL: cat coprocess 2 hung
/home/johno/GitRepos/KornShell/ksh/src/cmd/ksh93/tests/coprocess.sh: line 234: coprocess is running; cannot create a new coprocess
test coprocess(C.UTF-8) failed at 2024-01-27+18:14:31 with exit code 1 [ 35 tests 1 error ]
test io(C.UTF-8) begins at 2024-01-27+18:14:31
	io.sh[353]: FAIL: read -n3 from fifo failed -- expected 'a', got 'abc'
	io.sh[356]: FAIL: read -n1 from fifo failed -- expected 'b', got 'd'
test io(C.UTF-8) failed at 2024-01-27+18:14:34 with exit code 2 [ 166 tests 2 errors ]
test pty(C.UTF-8) begins at 2024-01-27+18:14:34
kill: 68177: no such process
	pty.sh[1027]: FAIL: suspend a blocked write to a FIFO: line 1040: expected "^\^C.*: testfifo: cannot create \[.*\]\r\n$", got "^C"
	pty.sh[1027]: FAIL: suspend a blocked write to a FIFO: line 1041: read timeout
<hangs unless pty is killed with 'pkill -x pty' or equivalent>
test pty(C.UTF-8) failed at 2024-01-27+18:15:39 with exit code 2 [ 61 tests 2 errors ]
Total errors: 5
```

- ~Use the libast signal function if `sigaction` is present (`_ast_signal`)~ Rename the libast `signal` function to `_ast_signal`.
- Add necessary sig.h include to map.c (the Mamfile was also updated to account for this).

Progresses https://github.com/ksh93/ksh/issues/518

JohnoKing committed:
Fix multiple test failures under ASan by switching to sigaction(2)

This commit fixes multiple test failures when the coprocess, io and pty
tests are run under ASan. Unlike the ksh93v- approach of implementing
sigaction via a sh_signal function, this commit repurposes the
(otherwise unused) signal wrapper function in libast, which already uses
sigaction for signal handling.

- Use the libast signal function if sigaction is present (_ast_signal).
- Add necessary sig.h include to map.c (the Mamfile was also updated
  to account for this).

JohnoKing committed:
Increment date

JohnoKing commented:
Actually, my previous explanation is incorrect. The actual reason this pull request works is not because ksh never uses `sigaction` (under normal circumstances it does use the `signal` wrapper). Rather, it fixes the failures under ASan because the `signal` wrapper causes a name conflict, similar to how the AST `glob` functions did before being fixed in https://github.com/ksh93/ksh/pull/282. Renaming the wrapper function to `_ast_signal` allows ksh to use `sigaction` under ASan without encountering a name conflict.

JohnoKing committed:
Merge with upstream/dev

McDutchie commented on src/lib/libast/misc/signal.c:
```diff
@@ -60,8 +60,8 @@
-extern Sig_handler_t
-signal(int sig, Sig_handler_t fun)
+Sig_handler_t
+_ast_signal(int sig, Sig_handler_t fun)
```
I don't think this rename is correct. The `#define signal _ast_signal` in map.c is emitted conditionally. If `signal` is *not* redefined to `_ast_signal`, then this rename should break things. If `signal` *is* redefined to `_ast_signal`, then this rename is unnecessary because the function will be implicitly renamed by the macro expansion. I could be missing something but this is how I think it works.

Oh, wait, I see the '#define signal' and '#undef signal' at the top. It should be possible to delete these (as well as undoing the rename) if that logic is transferred to map.c -- i.e., simply change line 138 in map.c to: `#if !__HAIKU__`.

I could be missing something and unfortunately I still don't have access to my x86_64 VMs right now, so I hope you have the time to test this.

JohnoKing replied:
The latest changes I applied have been successfully tested on Linux, Haiku, *BSD, illumos and Cygwin. (The `!__HAIKU` block was deleted as it's no longer necessary with the `_ast_signal` definition in map.c.)

JohnoKing committed:
Experimental commit with changes

I will test this commit on my various VMs after pushing. The rename in
signal.c was put there after I saw implicit-function warnings, but that
appears to have actually been caused by the undefs I overlooked.
In any case I don't *think* Haiku needs an extra if directive in map.c
because IIRC signal() is just a macro in the Haiku headers (to be
confirmed or deconfirmed with imminent testing).

McDutchie commented on src/lib/libast/features/sig.sh:
```diff
@@ -105,4 +105,5 @@ extern int		sigflag(int, int, int);
 extern int		sigcritical(int);
-extern int		sigunblock(int);'
+extern int		sigunblock(int);
+extern Sig_handler_t	_ast_signal(int, Sig_handler_t);'
```
This still has `_ast_signal`. Probably not needed now (?)

JohnoKing replied:
It is still required to avoid implicit function warnings on Haiku (due to the unique situation with `signal` being a macro in the OS headers).

JohnoKing committed:
Merge with upstream Mamfile changes
